component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/handlers/MTLSTokenBindingAuthorizationCodeGrantHandler.java

```diff
@@ -0,0 +1,139 @@
+                .getParameter(MutualTLSUtil.AUTHENTICATOR_TYPE_PARAM);
+        if (certHeader.isPresent() && MutualTLSUtil.AUTHENTICATOR_TYPE_MTLS.equals(authenticatorType)) {
+            Base64URL certThumbprint;
+            if (log.isDebugEnabled()) {
+                log.debug("Client MTLS certificate found: " + certHeader);
+            }
+            try {
+                X509Certificate certificate = parseCertificate(certHeader.get().getValue()[0]);
+                certThumbprint = X509CertUtils.computeSHA256Thumbprint(certificate);
+            } catch (CertificateException e) {
+                log.error("Error occurred while calculating the thumbprint of the client MTLS certificate", e);
+                return false;
+            }
+
+            // Add certificate thumbprint as a hidden scope of the token.
+            if (certThumbprint != null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Client MTLS certificate thumbprint: " + certThumbprint);
+                }
+                String[] scopes = tokReqMsgCtx.getScope();
+                List<String> scopesList = new LinkedList<>(Arrays.asList(scopes));
+                scopesList.add(MutualTLSUtil.CERT_THUMBPRINT + "#" + MutualTLSUtil.SHA256_DIGEST_ALGORITHM +
+                        MutualTLSUtil.CERT_THUMBPRINT_SEPARATOR + certThumbprint.toString());
+                tokReqMsgCtx.setScope(scopesList.toArray(new String[scopesList.size()]));
+            }
+        }
+        return validateScope;
+    }
+
+    /**
+     * Return Certificate for give Certificate Content.
+     *
+     * @param content Certificate Content.
+     * @return X509Certificate.
+     * @throws CertificateException Certificate Exception.
+     */
+    private static X509Certificate parseCertificate(String content) throws CertificateException {
+
+        // Trim extra spaces.
+        String decodedContent = content.trim();
+
+        // Remove certificate headers.
+        byte[] decoded = Base64.getDecoder().decode(decodedContent
+                .replaceAll(MutualTLSUtil.BEGIN_CERT, StringUtils.EMPTY)
+                .replaceAll(MutualTLSUtil.END_CERT, StringUtils.EMPTY).trim()
+        );
+
+        return (X509Certificate) CertificateFactory.getInstance("X.509")
+                .generateCertificate(new ByteArrayInputStream(decoded));
+    }
+
+    /**
+     * Remove the certificate thumbprint prefixed scope from the space delimited list of authorized scopes.
+     *
+     * @param scopes Authorized scopes of the token.
+     * @return scopes by removing the custom scope.
+     */
+    private String[] getReducedResponseScopes(String[] scopes) {
+        if (scopes != null && scopes.length > 0) {
+            List<String> scopesList = new LinkedList<>(Arrays.asList(scopes));
+            scopesList.removeIf(s -> s.startsWith(MutualTLSUtil.CERT_THUMBPRINT));
```

Comment:
use meaningful variable name instead of s

Reply:
Fixed cb61d22